// @faremeter/types/dist/src/x402.js
import { type } from "arktype";
export const x402PaymentRequirements = type({
    scheme: "string",
    network: "string",
    maxAmountRequired: "string.numeric",
    resource: "string.url",
    description: "string",
    mimeType: "string",
    outputSchema: "object?",
    payTo: "string",
    maxTimeoutSeconds: "number.integer",
    asset: "string",
    extra: "object?",
});
export const x402PaymentRequiredResponse = type({
    x402Version: "number.integer",
    accepts: x402PaymentRequirements.array(),
    error: "string?",
});
export const x402PaymentPayload = type({
    x402Version: "number.integer",
    scheme: "string",
    network: "string",
    asset: "string?",
    payload: "object",
});
export const x402PaymentHeaderToPayload = type("string.base64")
    .pipe.try((x) => atob(x))
    .to("string.json.parse")
    .to(x402PaymentPayload);
export const x402VerifyRequest = type({
    x402Version: "number.integer",
    paymentHeader: "string",
    paymentRequirements: x402PaymentRequirements,
});
export const x402VerifyResponse = type({
    isValid: "boolean",
    "invalidReason?": "string | null",
});
export const x402SettleRequest = x402VerifyRequest;
export const x402SettleResponse = type({
    success: "boolean",
    "error?": "string | null",
    txHash: "string | null",
    networkId: "string | null",
});
export const x402SupportedKind = type({
    x402Version: "number.integer",
    scheme: "string",
    network: "string",
    extra: "object?",
});
export const x402SupportedResponse = type({
    kinds: x402SupportedKind.array(),
});
