// @faremeter/info/dist/src/solana.js
import { addX402PaymentRequirementDefaults } from "./common.js";
import { Base58Address } from "@faremeter/types/solana";
const knownClusters = ["devnet", "testnet", "mainnet-beta"];
export function isKnownCluster(c) {
    return knownClusters.includes(c);
}
export const lookupX402Network = (network) => {
    const networks = [`solana-${network}`];
    if (network === "mainnet-beta") {
        networks.push("solana");
    }
    return networks;
};
const knownSPLTokens = {
    USDC: {
        cluster: {
            "mainnet-beta": {
                address: "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v",
            },
            devnet: {
                address: "4zMMC9srt5Ri5X14GAgXhaHii3GnPAEERYPJgZJDncDU",
            },
        },
        toUnit: (v) => v.toString(),
    },
};
export function lookupKnownSPLToken(cluster, name) {
    const splTokenInfo = knownSPLTokens[name];
    if (!splTokenInfo) {
        return;
    }
    const networkInfo = splTokenInfo.cluster[cluster];
    if (!networkInfo) {
        return;
    }
    return {
        ...networkInfo,
        cluster,
        name,
        toUnit: splTokenInfo.toUnit,
    };
}
export function isKnownSPLToken(splToken) {
    return splToken in knownSPLTokens;
}
export function x402Exact(args) {
    const tokenInfo = lookupKnownSPLToken(args.network, args.asset);
    if (!tokenInfo) {
        throw new Error(`couldn't look up token '${args.asset}' on Solana cluster`);
    }
    const networks = lookupX402Network(args.network);
    const req = networks.map((network) => addX402PaymentRequirementDefaults({
        scheme: "exact",
        network,
        maxAmountRequired: tokenInfo.toUnit(args.amount),
        payTo: args.payTo,
        asset: tokenInfo.address,
        maxTimeoutSeconds: 60, // from coinbase/x402's middleware defaults
    }));
    return req;
}
export function xSolanaSettlement(args) {
    let tokenInfo;
    // Special-case this, because it's not an SPL Token.
    if (args.asset === "sol") {
        tokenInfo = {
            address: "sol",
            toUnit: (x) => x.toString(),
        };
    }
    else {
        tokenInfo = lookupKnownSPLToken(args.network, args.asset);
    }
    if (!tokenInfo) {
        throw new Error(`couldn't look up token '${args.asset}' on Solana cluster`);
    }
    const req = addX402PaymentRequirementDefaults({
        scheme: "@faremeter/x-solana-settlement",
        network: args.network,
        maxAmountRequired: tokenInfo.toUnit(args.amount),
        payTo: args.payTo,
        asset: tokenInfo.address,
        maxTimeoutSeconds: 60, // from coinbase/x402's middleware defaults
    });
    return req;
}
